### DISCUSSION TITLE
How to use the Jotai Store API?

### DISCUSSION THREAD

**Discussion by rodrigocfd** [pmndrs/jotai#1752] [Q&A] [answered]
Jotai v2 features a new [Store API](https://jotai.org/docs/guides/migrating-to-v2-api#store-api), but the project [README](https://github.com/pmndrs/jotai/blob/main/readme.md) does not mention it. I'm wondering what's the best way to use.

My use case is a large enterprise application, but in a simplified view, I have global objects like these:

```ts
import {atom} from 'jotai';

const warningsAtom = atom([] as string[]);
const errorsAtom = atom([] as string[]);
```

And a component, at the top of the page, which displays them:

```ts
import {useAtom} from 'jotai';
import {errorsAtom, warningsAtom} from './atoms';

function GlobalMessages() {
	const [warnings] = useAtom(warningsAtom);
	const [errors] = useAtom(errorsAtom);

	return <>
		{warnings.length > 0 && 
			<div>
				<h2>Warnings</h2>
				{warnings.map((w, i) =>
					<div key={i}>{w}</div>
				)}
			</div>
		}

		{errors.length > 0 &&
			<div>
				<h2>Errors</h2>
				{errors.map((e, i) =>
					<div key={i}>{e}</div>
				)}
			</div>
		}
	</>;
}
```

Now some components can send/receive server data, according to user actions. There is a `serverPost()` function which fetches data from the back-end, and depending on the returned data, can add messages to `warnings` or `errors`.

Then it comes the problem: `serverPost()` must access the global object to write to `warnings` and `errors`.

As far as I could see, now I can have something like:

```ts
import {createStore} from 'jotai';
import {warningsAtom} from './atoms';

async function serverPost<T>(url: string): Promise<T> {
	
	/// fetch() ...
	
	const store = createStore();
	const warnings = store.get(warningsAtom);
	store.set(warningsAtom, [...warnings, 'New warning']);
	
}
```

Is this correct?

**Comment by dai-shi** [comment]
Before diving into the specific care, there's something:

While store api is exposed and stable in a sense it's really used in the code (src/react),
we still consider the best practice is unknown.
So, I'd say it's for experienced jotai users currently, and we are looking for best practice.

In general, here's my guideline as of writing.
- We shouldn't use store api as much as possible. Using store directly means, it's less React-ish. That was the one of huge reasons it's not exposed in v1.
- If you want to access only a few atom values outside react, using useEffect or syncing atom would be preferable.
- Library authors who use jotai internally must use `createStore` and custom contexts to avoid conflicting.

Other than that, I'm not sure. Looking forward to finding creative usage of store api.

**Comment by dai-shi** [comment]
> ```ts
> import {createStore} from 'jotai';
> import {warningsAtom} from './atoms';
> 
> async function serverPost<T>(url: string): Promise<T> {
> 	
> 	/// fetch() ...
> 	
> 	const store = createStore();
> 	const warnings = store.get(warningsAtom);
> 	store.set(warningsAtom, [...warnings, 'New warning']);
> 	
> }
> ```
> 
> Is this correct?

It looks wrong. When you create a new store, you should almost always pass it with `<Provider store={store}>` (there are some exceptions).

**Comment by rodrigocfd** [comment, marked as answer]
Can you write up an example of how this situation could be implemented?

**Reply by dai-shi** [reply]
I wouldn't probably use `createStore` for this example, but if you do, please define it at the module and pass it to `Provider` in your root component.

```tsx
const store = createStore()

async function serverPost<T>(url: string): Promise<T> {
  // fetch() ...
  store.set(warningsAtom, prev => [...prev, 'New warning']);
}

const App = () => (
  <Provider store={store}>
    ...
  </Provider>
)
```
